### PR TITLE
allow multiple factions on a card and finish implementing ward

### PR DESCRIPTION
**Pull request by cryogen**

### CODE REVIEW

File: server/game/basecard.js [hunk 1]
```diff
@@ -324,6 +327,10 @@ class BaseCard {
         return this.cardData.type_code;
     }
 
+    getPrintedFaction() {
+        return this.cardDate.faction_code;
```
Comment (ystros) [inline server/game/basecard.js]: Typo - cardData.

File: server/game/basecard.js [hunk 2]
```diff
@@ -247,6 +250,10 @@ class BaseCard {
         return !!this.traits[trait.toLowerCase()];
     }
 
+    isFaction(faction) {
```
Comment (ystros) [inline server/game/basecard.js]: I think `hasFaction` would be a more consistent name with other methods we have (hasTrait, hasKeyword, etc).

File: server/game/cards/characters/04/shadowtowermason.js [hunk 3]
```diff
@@ -14,7 +14,7 @@ class ShadowTowerMason extends DrawCard {
 
     getCardCount() {
         var count = this.controller.allCards.reduce((counter, card) => {
-            if(this.isBlank() || card.location !== 'play area' || card.getFaction() !== 'thenightswatch' || (card.getType() !== 'location' && card.getType() !== 'attachment')) {
+            if(this.isBlank() || card.location !== 'play area' || card.isFaction('thenightswatch' || (card.getType() !== 'location' && card.getType() !== 'attachment'))) {
```
Comment (ystros) [inline server/game/cards/characters/04/shadowtowermason.js]: Missing or misplaced close paren on the faction check.

File: server/game/cards/characters/05/serlancellannister.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ class SerLancelLannister extends DrawCard {
 
     getSingleOtherLannisterLordOrLady() {
         var cards = this.controller.cardsInPlay.filter(card => {
-            return card.getFaction() === 'lannister' && (card.hasTrait('Lord') || card.hasTrait('Lady')) && card.getType() === 'character' && card !== this;
+            return card.isFaction('lannister' && (card.hasTrait('Lord') || card.hasTrait('Lady')) && card.getType() === 'character') && card !== this;
```
Comment (ystros) [inline server/game/cards/characters/05/serlancellannister.js]: Missing close paren.

File: server/game/cards/locations/04/harrenhalffh.js [hunk 5]
```diff
@@ -12,7 +12,7 @@ class Harrenhal extends DrawCard {
     kneel(player) {
         this.game.promptForSelect(player, {
             cardCondition: card => card.location === 'hand' && card.getType() === 'character' && card.controller === this.controller &&
-                (card.getFaction() === 'lannister' || card.hasTrait('House Bolton')),
+                (card.isFaction('lannister' || card.hasTrait('House Bolton'))),
```
Comment (ystros) [inline server/game/cards/locations/04/harrenhalffh.js]: Close paren

File: server/game/basecard.js [hunk 6]
```diff
@@ -52,6 +52,9 @@ class BaseCard {
         this.parseKeywords(cardData.text || '');
         this.parseTraits(cardData.traits || '');
         this.setupCardAbilities(AbilityDsl);
+
+        this.factions = {};
+        this.factions[cardData.faction_code] = 1;
```
Comment (ystros) [inline server/game/basecard.js]: Should toLowerCase the faction here for consistency. Or better yet, use the addFaction method.

File: server/game/basecard.js (hunk 2 above)
Comment (cryogen) [inline server/game/basecard.js]: as discussed on discord, i think it reads more naturally this way

**Pull request merged**